### ISSUE TITLE
We should have a default hasher for Strings for CMS

### ISSUE THREAD

**Issue by johnynek** [twitter/algebird#398]
Here are two approaches:

1) hash using Murmur to Long, then use the Long universal hash functions.

2) append the a and b we get for the family of hashes and use Murmur on the whole block.

We could also add `Array[Byte]` while we are at it since we will probably go through that anyway.

**Comment by miguno** [comment]
That's a good suggestion.  I have been thinking about the very same as we have been using `CMS[BigInt]` to effectively do `CMS[String]` (you can easily convert BigInt<->String with built-in methods). :-)

> 1) hash using Murmur to Long, then use the Long universal hash functions.

Hmm.  Why would you want to hash twice, i.e. `hashLong(hashMurmur(s))` to get the desired `Int`?

Murmur already gives us the desired `Int`, hence doing only `hashMurmur(s)` is faster.  Furthermore, I *think* the Murmur-only approach would provide us with a more uniform distribution of `Int`s then a double-hash approach;  at least in the Murmur-only case we *know* based on prior work (I don't mean #392 necessarily, I mean studies of Murmur3 in general) that the `Int` output is well-formed, whereas we'd have to do our own research/plotting/etc. to understand the impact of the proposed double-hashing.  For instance, `(ax + b) mod p` is said to sort-of [produce predictable patterns](http://stackoverflow.com/a/12120625/1743580) -- to me this sounds like it would lower the quality of the Murmur hashing, and we'd have to do the latter anyways.

Similarly, my experience with #392 makes me feel that we should rely on just Murmur3, as the `(ax + b) mod p` hashing resulted in counter-intuitive (read: wrong, IMHO) behavior when hashing `BigInt`s.  And as I said above, we have been using `CMS[BigInt]` to effectively do `CMS[String]`, so I can claim limited practical experience with it. :-)

> 2) append the a and b we get for the family of hashes and use Murmur on the whole block.

Not sure I understand -- why would you want to append `a` and/or `b` to the string before hashing?

> We could also add Array[Byte] while we are at it since we will probably go through that anyway.

The `BigInt` hashing is effectively hashing `Array[Byte]` already, so that would be a starting point.  

Here's a snippet for hashing `BigInt`, `String`, `Array[Byte]`:

```scala
  implicit object CMSHasherArrayByte extends CMSHasher[Array[Byte]] {
    override def hash(a: Int, b: Int, width: Int)(x: Array[Byte]): Int = {
      val hash: Int = scala.util.hashing.MurmurHash3.arrayHash(x, a)
      val positiveHash = hash & Int.MaxValue
      positiveHash % width
    }
  }

  implicit object CMSHasherBigInt extends CMSHasher[BigInt] {
    override def hash(a: Int, b: Int, width: Int)(x: BigInt): Int =
        CMSHasherArrayByte.hash(a, b, width)(x.toByteArray)
  }

  implicit object CMSHasherString extends CMSHasher[String] {
    override def hash(a: Int, b: Int, width: Int)(x: String): Int =
        CMSHasherArrayByte.hash(a, b, width)(x.getBytes("UTF-8"))
  }
```

Given my (limited) research into this hashing topic over the past few days I think the snippet above would address the `String` hashing in a simple as well as effective way.

The one open question I still have is what the memory and GC impacts of the Murmur-based setup are in practice.  I've only done limited testing so far, though this should change over the next couple of days.

Just my two cents! :-)

PS: There is also a section on [hashing strings](http://en.wikipedia.org/wiki/Universal_hashing#Hashing_strings) in the Wikipedia article on [Universal Hashing](http://en.wikipedia.org/wiki/Universal_hashing#Hashing_strings).

**Comment by johnynek** [comment]
The ones you posted here seem good to me.

I wonder if we could cleanly make a contramap method on the count-min-sketch monoid so that we could take a function from `L => K` on a `CMS[K]` monoid to produce `CMS[L]` monoid. This might make it easy to work with, say, the `CMS[String]` monoid and use `(_.toString)` as a cheap function to get a quick and dirty CMS.

**Comment by miguno** [comment]
I like the contramap idea!  I sent a pull request with a first draft.

**Comment by miguno** [comment]
This ticket can be closed as this functionality is covered in https://github.com/twitter/algebird/pull/399, which will be included in upcoming Algebird 0.10.
